WcaOnRails/app/controllers/delegate_reports_controller.rb

```diff
@@ -0,0 +1,61 @@
+class DelegateReportsController < ApplicationController
+  before_action :authenticate_user!
+
+  before_action -> { redirect_unless_user(:can_view_delegate_report?, competition_from_params.delegate_report) }, only: [
+    :show,
+  ]
+  before_action -> { redirect_unless_user(:can_edit_delegate_report?, competition_from_params.delegate_report) }, only: [
+    :edit,
+    :update,
+  ]
```

Comment:
I like `before_action :authenticate_user!` because it's general for all actions. Current, and most likely future ones.

Don't the others because (1) they only apply to enumerated actions, making that code harder to read, and (2) `competition_from_params.delegate_report` is computed twice.